### PR TITLE
fix: fix for opcode "invoke-custom"'s referenceType CALL_SITE

### PR DESCRIPTION
**Pull request by farfromrefug**
I am not a technical expert on smali.
The solution comes from @MG1937 [here](https://github.com/google/smali/issues/66#issuecomment-2354362764). I applied his change and it fixed https://github.com/google/smali/issues/66 Hope this is enough to get this PR to be merged

### CODE REVIEW

File: dexlib2/src/main/java/com/android/tools/smali/dexlib2/writer/DexWriter.java [hunk 1]
```diff
@@ -1256,6 +1256,8 @@ private int writeCodeItem(@Nonnull DexDataWriter writer,
                     if (paramCount > outParamCount) {
                         outParamCount = paramCount;
                     }
+                } else if (instruction.getOpcode().referenceType == ReferenceType.CALL_SITE) {
+                    outParamCount = ((VariableRegisterInstruction)instruction).getRegisterCount();
```
Comment (sgjesse) [inline dexlib2/src/main/java/com/android/tools/smali/dexlib2/writer/DexWriter.java:1260]: Judging from the code above, I think this should be
```
                    int paramCount = ((VariableRegisterInstruction)instruction).getRegisterCount();
                    if (paramCount > outParamCount) {
                        outParamCount = paramCount;
                    }
```
I suggest moving the declaration of `paramCount` one level out and share the update to `outParamCount`. A test should be able to reveal how this should be.

**Review by sgjesse — COMMENTED** [review]
Thank you for the pull request. I have a comment, and please add a few tests demonstrating the issue and fix.

**Comment by IgorEisberg** [comment]
For extensibility's sake I suggest this:
```
                int paramCount = 0;
                switch (instruction.getOpcode().referenceType) {
                    case METHOD: {
                        ReferenceInstruction refInsn = (ReferenceInstruction)instruction;
                        MethodReference methodRef = (MethodReference)refInsn.getReference();
                        Opcode opcode = instruction.getOpcode();
                        if (InstructionUtil.isInvokePolymorphic(opcode)) {
                            paramCount = ((VariableRegisterInstruction)instruction).getRegisterCount();
                        } else {
                            paramCount = MethodUtil.getParameterRegisterCount(methodRef, InstructionUtil.isInvokeStatic(opcode));
                        }
                        break;
                    }
                    case CALL_SITE:
                        paramCount = ((VariableRegisterInstruction)instruction).getRegisterCount();
                        break;
                }
                if (paramCount > outParamCount) {
                    outParamCount = paramCount;
                }
```

**Comment by farfromrefug** [comment]
@sgjesse i modified as per @IgorEisberg comment as it was also taking your comment into account. Hope it is good.
as for the tests, i really dont know how to write them :s I created that PR based on someone comment (who did not want to make a PR). I dont really know much about smali basecode.
Hope it is ok with you!

**Comment by IgorEisberg** [comment]
> @sgjesse i modified as per @IgorEisberg comment as it was also taking your comment into account. Hope it is good. as for the tests, i really dont know how to write them :s I created that PR based on someone comment (who did not want to make a PR). I dont really know much about smali basecode. Hope it is ok with you!

Not sure how you're supposed to write a test for something that is invisible when baksmaling a broken rebuilt dex and requires ART runtime to verify... With that said, I can vouch that it works properly, and I rebuilt a whole Android 15 ROM with it. No more VerifyError.
The easiest way to test it without a runtime is using `dexdump -d classes.dex`.
Notice the differences in `outs`.

Broken dex output:
```    #2              : (in Lcom/android/server/art/DexMetadataHelper$DexMetadataInfo;)
      name          : 'equals'
      type          : '(Ljava/lang/Object;)Z'
      access        : 0x0011 (PUBLIC FINAL)
      code          -
      registers     : 2
      ins           : 2
      outs          : 0
      insns size    : 5 16-bit code units
040a08:                                        |[040a08] com.android.server.art.DexMetadataHelper$DexMetadataInfo.equals:(Ljava/lang/Object;)Z
040a18: fc20 0000 1000                         |0000: invoke-custom {v0, v1}, call_site@0000
040a1e: 0a00                                   |0003: move-result v0
040a20: 0f00                                   |0004: return v0
      catches       : (none)
      positions     : 
      locals        : 

    #3              : (in Lcom/android/server/art/DexMetadataHelper$DexMetadataInfo;)
      name          : 'hashCode'
      type          : '()I'
      access        : 0x0011 (PUBLIC FINAL)
      code          -
      registers     : 1
      ins           : 1
      outs          : 0
      insns size    : 5 16-bit code units
040a24:                                        |[040a24] com.android.server.art.DexMetadataHelper$DexMetadataInfo.hashCode:()I
040a34: fc10 0200 0000                         |0000: invoke-custom {v0}, call_site@0002
040a3a: 0a00                                   |0003: move-result v0
040a3c: 0f00                                   |0004: return v0
      catches       : (none)
      positions     : 
      locals        : 

    #4              : (in Lcom/android/server/art/DexMetadataHelper$DexMetadataInfo;)
      name          : 'toString'
      type          : '()Ljava/lang/String;'
      access        : 0x0011 (PUBLIC FINAL)
      code          -
      registers     : 1
      ins           : 1
      outs          : 0
      insns size    : 5 16-bit code units
040a40:                                        |[040a40] com.android.server.art.DexMetadataHelper$DexMetadataInfo.toString:()Ljava/lang/String;
040a50: fc10 0100 0000                         |0000: invoke-custom {v0}, call_site@0001
040a56: 0c00                                   |0003: move-result-object v0
040a58: 1100                                   |0004: return-object v0
      catches       : (none)
      positions     : 
      locals        : 
```

Fixed dex output:
```
    #2              : (in Lcom/android/server/art/DexMetadataHelper$DexMetadataInfo;)
      name          : 'equals'
      type          : '(Ljava/lang/Object;)Z'
      access        : 0x0011 (PUBLIC FINAL)
      code          -
      registers     : 2
      ins           : 2
      outs          : 2
      insns size    : 5 16-bit code units
040a08:                                        |[040a08] com.android.server.art.DexMetadataHelper$DexMetadataInfo.equals:(Ljava/lang/Object;)Z
040a18: fc20 0000 1000                         |0000: invoke-custom {v0, v1}, call_site@0000
040a1e: 0a00                                   |0003: move-result v0
040a20: 0f00                                   |0004: return v0
      catches       : (none)
      positions     : 
      locals        : 

    #3              : (in Lcom/android/server/art/DexMetadataHelper$DexMetadataInfo;)
      name          : 'hashCode'
      type          : '()I'
      access        : 0x0011 (PUBLIC FINAL)
      code          -
      registers     : 1
      ins           : 1
      outs          : 1
      insns size    : 5 16-bit code units
040a24:                                        |[040a24] com.android.server.art.DexMetadataHelper$DexMetadataInfo.hashCode:()I
040a34: fc10 0200 0000                         |0000: invoke-custom {v0}, call_site@0002
040a3a: 0a00                                   |0003: move-result v0
040a3c: 0f00                                   |0004: return v0
      catches       : (none)
      positions     : 
      locals        : 

    #4              : (in Lcom/android/server/art/DexMetadataHelper$DexMetadataInfo;)
      name          : 'toString'
      type          : '()Ljava/lang/String;'
      access        : 0x0011 (PUBLIC FINAL)
      code          -
      registers     : 1
      ins           : 1
      outs          : 1
      insns size    : 5 16-bit code units
040a40:                                        |[040a40] com.android.server.art.DexMetadataHelper$DexMetadataInfo.toString:()Ljava/lang/String;
040a50: fc10 0100 0000                         |0000: invoke-custom {v0}, call_site@0001
040a56: 0c00                                   |0003: move-result-object v0
040a58: 1100                                   |0004: return-object v0
      catches       : (none)
      positions     : 
      locals        : 
```

**Review by Blgwp21 — APPROVED** [review]

**Review by iBotPeaches — APPROVED** [review]

**Review by Blgwp21 — APPROVED** [review]

**Comment by sgjesse** [comment]
OK, I will approve and land this.

**Review by sgjesse — APPROVED** [review]

**Comment by iBotPeaches** [comment]
smali patch release before end of year possible?

**Pull request merged**